### PR TITLE
added support to save state and paste it (normal blend) on top of the result

### PR DESCRIPTION
**Pull request by yosefBazaart**
add the capability to save the current state and paste it after a few filters are applied

### CODE REVIEW

File: library/src/main/java/jp/co/cyberagent/android/gpuimage/filter/GPUImageFilterGroup.kt [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+     * @param filters the filters which represent this filter
+     */
+    /**
+     * Instantiates a new GPUImageFilterGroup with no filters.
+     */
+    init {
+        updateMergedFilters()
+        glCubeBuffer = ByteBuffer.allocateDirect(GPUImageRenderer.CUBE.size * 4)
+            .order(ByteOrder.nativeOrder())
+            .asFloatBuffer()
```
Comment (NivBazaart) [inline library/src/main/java/jp/co/cyberagent/android/gpuimage/filter/GPUImageFilterGroup.kt:53]: '''suggestion
// Allocate a direct byte buffer for the cube vertices:
glCubeBuffer = ByteBuffer.allocateDirect(GPUImageRenderer.CUBE.size * 4)
.order(ByteOrder.nativeOrder())
.asFloatBuffer()

File: library/src/main/java/jp/co/cyberagent/android/gpuimage/filter/GPUImageFilterGroup.kt [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+     * Instantiates a new GPUImageFilterGroup with the given filters.
+     *
+     * @param filters the filters which represent this filter
+     */
+    /**
+     * Instantiates a new GPUImageFilterGroup with no filters.
+     */
+    init {
+        updateMergedFilters()
+        glCubeBuffer = ByteBuffer.allocateDirect(GPUImageRenderer.CUBE.size * 4)
```
Comment (NivBazaart) [inline library/src/main/java/jp/co/cyberagent/android/gpuimage/filter/GPUImageFilterGroup.kt:51]:
```suggestion
        // Allocate a direct byte buffer for the cube vertices:
        glCubeBuffer = ByteBuffer.allocateDirect(GPUImageRenderer.CUBE.size * 4)
```

File: library/src/main/java/jp/co/cyberagent/android/gpuimage/filter/GPUImageFilterGroup.kt [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+    fun clearFilters() {
+        filters.clear()
+    }
+
+    fun getMergedFilters(): List<GPUImageFilter>? {
+        return mergedFilters
+    }
+
+    fun updateMergedFilters() {
+        val mergedFilters = mergedFilters ?: ArrayList()
```
Comment (UrK) [inline library/src/main/java/jp/co/cyberagent/android/gpuimage/filter/GPUImageFilterGroup.kt:259]: Why assigning the old filters if the next line just clears the array?

File: library/src/main/java/jp/co/cyberagent/android/gpuimage/filter/GPUImageFilterGroup.kt [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+
+/**
+ * Resembles a filter that consists of multiple filters applied after each
+ * other.
+ */
+open class GPUImageFilterGroup @JvmOverloads constructor(private val filters: MutableList<GPUImageFilter> = mutableListOf()) :
+    GPUImageFilter() {
+
+    private lateinit var frameBufferIndexMap: Map<Int, Int>
+    private var mergedFilters: MutableList<GPUImageFilter>? = null
```
Comment (UrK) [inline library/src/main/java/jp/co/cyberagent/android/gpuimage/filter/GPUImageFilterGroup.kt:35]: `mergedFilters` is never null

Comment (UrK) [reply library/src/main/java/jp/co/cyberagent/android/gpuimage/filter/GPUImageFilterGroup.kt:35]: Wait... where is this code copied from?

**Review by UrK — APPROVED** [review]

**Pull request merged**